Add an annotation for hand-written type-guards `ts-auto-guard:custom`

webstrand opened the pull request:
`ts-auto-guard` can't generate correct type-guards for branded literal types:


```ts
/** @see {isPersonId} ts-auto-guard:type-guard */
export type PersonId = number & { brand: true };

// produces this:
export function isPersonId(obj: unknown): obj is PersonId {
    const typedObj = obj as PersonId
    return (
        typeof typedObj === "number" &&
        (typedObj !== null &&
            typeof typedObj === "object" ||
            typeof typedObj === "function") &&
        typedObj["brand"] === true
    )
}
```

Using this new annotation, auto-generation can be bypassed and a correct type-guard can be hand-written:

```ts
/** @see {isPersonId} ts-auto-guard:custom */
export type PersonId = number & { brand: true };
export function isPersonId(x: unknown): x is PersonId {
    return typeof x === "string"
}
```

Custom validators are useful for edge cases other than just branded literal types, where non-structural checks are desirable too. And it allows for integration with other type-guard generators like Zod.

I considered modifying the intersection logic to prune intersections between a literal and an object type, but that may quietly hide issues in the users types that would otherwise be surfaced by the type-guard rejecting all candidate values.

webstrand committed:
Fix phantom dependency on @ts-morph/common

depending on the package manager used, this was previously installed
as a phantom dependency

webstrand committed:
Add hand-written type-guard support with ts-auto-guard:custom

schmop commented:
ts-auto-guard is a tool to generate type-guards automatically.
Why exactly would you need an annotation to stop it doing that? 
Couldn't you just not annotate your type and typeguard, and then ts-auto-guard wouldn't override any self-written guards?

webstrand commented:
Ah yeah I should have been clearer about that in the pull request. I need a way to tell `ts-auto-guard` to use my hand-written type-guard when automatically generating guards for _other_ types.

schmop commented on package.json:
```diff
@@ -21,6 +21,7 @@
   "author": "Rhys van der Waerden",
   "license": "MIT",
   "dependencies": {
+    "@ts-morph/common": "^0.25.0",
```
Seems not to be needed, also @ts-morph/common says "Please do not depend on this library"
https://github.com/dsherret/ts-morph/blob/699815f54ae9b5c2a93f016ba1a9df1e8ac1c014/packages/common/readme.md

webstrand replied:
The issue is:

https://github.com/rhys-vdw/ts-auto-guard/blob/5f6266d8daff7608a8b0d653d9b2f7330d937404/src/index.ts#L3

uses `@ts-morph/common` even though its not declared as dependency. This makes the program fail at runtime when installed via non-hoisting package managers.

I could replace the use of it 

https://github.com/rhys-vdw/ts-auto-guard/blob/5f6266d8daff7608a8b0d653d9b2f7330d937404/src/index.ts#L785

with `process.cwd().replace(/\\/g, '/')` and remove the import entirely?

But this probably could be split out into another PR. I had to fix this for mysef so I could build for testing the library.

----
Example error using `ts-auto-guard`:

```bash
tmpdir="$(mktemp -d)"
(cd $tmpdir; pnpm init && pnpm add --no-hoist ts-auto-guard && node -e 'require("ts-auto-guard")')
# Error: Cannot find module '@ts-morph/common'
rm -r "$tmpdir"

```

And a similar error building ts-auto-guard itself using pnpm:

```bash
tmpdir="$(mktemp -d)"
git worktree add --detach "$tmpdir" '5f6266d8da^'
(cd "$tmpdir" && pnpm install && pnpm build)
# Error: Cannot find module '@ts-morph/common'
git worktree remove --force "$tmpdir"
```

schmop commented on README.md:
```diff
@@ -206,3 +206,27 @@ export function isPerson(obj: unknown): obj is Person {
+
+## Use Custom Type-Guard Instead of Generating
+
+ts-auto-guard cannot generate type-guards for all typescript types automatically. For instance a validator for string template literals or branded types cannot be automatically genetrated. If you want to use a type which cannot be validated automatically you can use the annotation `/** @see {name} ts-auto-guard:type-guard */`, where `name` is a function exported by the current file:
```
```suggestion
ts-auto-guard cannot generate type-guards for all typescript types automatically. For instance a validator for string template literals or branded types cannot be automatically genetrated. If you want to use a type which cannot be validated automatically you can use the annotation `/** @see {name} ts-auto-guard:custom */`, where `name` is a function exported by the current file:
```

schmop replied:
And can you add links to the topics/concepts like string template literals and branded types?

I searched myself for a good reference for branded types and did not find anything from the typescript-lang community. That is probably because they do not like nominal type features in the structural type system, that TypeScript decided about.

https://github.com/microsoft/TypeScript/issues/11643#issuecomment-254891404

https://github.com/microsoft/TypeScript/issues/202#issuecomment-50916699

webstrand replied:
I added references, good idea. On the typescript discord, they're pretty familiar with branded types, but the ecosystem lacks good documentation for a lot of these type-level tricks.

schmop commented on README.md:
```diff
@@ -206,3 +206,27 @@ export function isPerson(obj: unknown): obj is Person {
+/** @see {isPersonId} ts-auto-guard:custom */
+export type PersonId = number & { brand: true };
+
+export function isPersonId(x: unknown): x is string {
```
```suggestion
export function isPersonId(x: unknown): x is PersonId {
```

schmop reviewed (dismissed):
I really like this feature. Added some comments that need to be resolved, will try out later practice.

rhys-vdw commented:
Seems like a sensible feature given the limitations.

webstrand committed:
Fix documentation isPersonId return type

webstrand committed:
Rewrite ts-auto-guard:custom documentation

webstrand committed:
Remove dependency on @ts-morph/common

webstrand commented:
I rewrote the documentation substantially and added references for branded/nominal types and for template literal strings. I also split out three common use cases I see for the feature with examples. I hope this improves things.

I'm quite happy to see this might be getting merged, I prefer ts-auto-guard to typia.

schmop commented:
Very nice rework on the Documentation!

LGTM, @rhys-vdw you good with merging?

rhys-vdw commented:
> I'm quite happy to see this might be getting merged, I prefer ts-auto-guard to typia.

I was going to ask. 😆

rhys-vdw commented on src/index.ts:
```diff
@@ -782,7 +784,7 @@ function propertyConditions(
   )
   if (debug) {
     if (expectedType.indexOf('import') > -1) {
-      const standardizedCwd = FileUtils.standardizeSlashes(process.cwd())
```
What's wrong with `standardizeSlashes`?

rhys-vdw replied:
oh sorry this was covered above. Don't mind me!

rhys-vdw reviewed (changes requested):
Just one note to address.

rhys-vdw commented:
Looks good, @webstrand. I've invited you as a collaborator if you're interested: https://github.com/rhys-vdw/ts-auto-guard/invitations

Feel free to join if you like. Just pls raise issues first and get someone else to review your work before merging.

rhys-vdw commented:
@webstrand lol... had to update the release pipeline and screwed it up a few times, but this feature is now live in https://github.com/rhys-vdw/ts-auto-guard/releases/tag/v5.1.3
